File: packages/runtime/src/_pdv/descriptor.py
"""Wire-shape dataclasses.

These mirror ``packages/protocol/src/descriptor.ts`` one-to-one. Python uses
snake_case internally and the wire uses camelCase, so each class owns an
explicit ``to_dict``. The duplication is deliberate: it is the one place where a
protocol change forces a visible edit on the Python side, which is exactly what
you want at a version boundary.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

# Element types that may go on the wire. Anything else must be converted by the
# adapter, with the original recorded in `Descriptor.dtype`.
WIRE_DTYPES = frozenset(
    {"f32", "f64", "i8", "i16", "i32", "i64", "u8", "u16", "u32", "u64", "bool"}
)

BYTES_PER_ELEMENT = {
    "f32": 4,
    "f64": 8,
    "i8": 1,
    "i16": 2,
    "i32": 4,
    "i64": 8,
    "u8": 1,
    "u16": 2,
    "u32": 4,
    "u64": 8,
    "bool": 1,
}


@dataclass
class NumericStats:
    """Summary statistics over the *complete* value, never over a subsample.

    See the note in the TypeScript mirror: decimation must not change these
    numbers, or the stats strip starts quietly lying about the data.
    """

    count: int
    min: Optional[float] = None
    max: Optional[float] = None
    mean: Optional[float] = None
    std: Optional[float] = None
    nan_count: int = 0
    inf_count: int = 0

    def to_dict(self) -> Dict[str, Any]:
        return {
            "count": int(self.count),
            "min": _finite_or_none(self.min),
            "max": _finite_or_none(self.max),
            "mean": _finite_or_none(self.mean),
            "std": _finite_or_none(self.std),
            "nanCount": int(self.nan_count),
            "infCount": int(self.inf_count),
        }


@dataclass
class Channel:
    """One contiguous run of numbers inside the payload buffer."""

    name: str
    role: str
    dtype: str
    length: int
    byte_offset: int
    byte_length: int
    stats: Optional[NumericStats] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "role": self.role,
            "dtype": self.dtype,
            "length": int(self.length),
            "byteOffset": int(self.byte_offset),
            "byteLength": int(self.byte_length),
            "stats": self.stats.to_dict() if self.stats else None,
        }


@dataclass
class Decimation:
    method: str
    original_length: int
    output_length: int

    def to_dict(self) -> Dict[str, Any]:
        return {
            "method": self.method,
            "originalLength": int(self.original_length),
            "outputLength": int(self.output_length),
        }


@dataclass
class IndexInfo:
    kind: str
    name: Optional[str] = None
    dtype: Optional[str] = None
    channel: Optional[str] = None
    time_unit: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "kind": self.kind,
            "name": self.name,
            "dtype": self.dtype,
            "channel": self.channel,
            "timeUnit": self.time_unit,
        }


@dataclass
class ColumnInfo:
    name: str
    dtype: str
    numeric: bool
    channel: Optional[str] = None
    stats: Optional[NumericStats] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "dtype": self.dtype,
            "numeric": bool(self.numeric),
            "channel": self.channel,
            "stats": self.stats.to_dict() if self.stats else None,
        }


@dataclass
class Descriptor:
    kind: str
    python_type: str
    preview: str
    shape: Optional[List[int]] = None
    dtype: Optional[str] = None
    nbytes: Optional[int] = None
    stats: Optional[NumericStats] = None
    index: Optional[IndexInfo] = None
    columns: Optional[List[ColumnInfo]] = None
    channels: List[Channel] = field(default_factory=list)
    decimation: Optional[Decimation] = None
    truncated: bool = False
    suggested_viz: List[str] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "kind": self.kind,
            "pythonType": self.python_type,
            "preview": self.preview,
            "shape": [int(d) for d in self.shape] if self.shape is not None else None,
            "dtype": self.dtype,
            "nbytes": int(self.nbytes) if self.nbytes is not None else None,
            "stats": self.stats.to_dict() if self.stats else None,
            "index": self.index.to_dict() if self.index else None,
            "columns": [c.to_dict() for c in self.columns] if self.columns is not None else None,
            "channels": [c.to_dict() for c in self.channels],
            "decimation": self.decimation.to_dict() if self.decimation else None,
            "truncated": bool(self.truncated),
            "suggestedViz": list(self.suggested_viz),
        }


@dataclass
class Capture:
    """An adapter's complete answer: metadata plus the raw bytes it references."""

    descriptor: Descriptor
    payload: bytes = b""
    warnings: List[str] = field(default_factory=list)


def _finite_or_none(value: Optional[float]) -> Optional[float]:
    """Keep NaN and Inf out of the JSON, which cannot represent them.

    Note this only guards *statistics*. Actual data values keep their NaN and
    Inf bit patterns, because they travel as raw binary rather than as JSON.
    """
    if value is None:
        return None
    value = float(value)
    if value != value or value in (float("inf"), float("-inf")):
        return None
    return value

File: packages/protocol/src/descriptor.ts
import { z } from "zod";
import { wireDtypeSchema } from "./dtype";

/** Bumped whenever the shape of a capture response changes incompatibly. */
export const PROTOCOL_VERSION = 1;

/**
 * Summary statistics for a numeric channel or column.
 *
 * Invariant, and the reason this type exists at all: these are always computed
 * over the *complete* value, before any decimation or truncation. A plot may
 * show 2000 of 4 million points, but the min/max/NaN counts shown next to it
 * still describe all 4 million. Silently reporting statistics of a subsample is
 * the single easiest way for a debugging tool to mislead someone.
 */
export const numericStatsSchema = z.object({
  /** Number of elements considered, including NaN and Inf. */
  count: z.number().int().nonnegative(),
  /** Extremes over finite values only; null when there are none. */
  min: z.number().nullable(),
  max: z.number().nullable(),
  mean: z.number().nullable(),
  std: z.number().nullable(),
  nanCount: z.number().int().nonnegative(),
  infCount: z.number().int().nonnegative(),
});
export type NumericStats = z.infer<typeof numericStatsSchema>;

/**
 * What a channel is for. The webview uses this to wire channels up to a
 * visualization without having to guess from names.
 */
export const channelRoleSchema = z.enum([
  "x",
  "y",
  "value",
  "open",
  "high",
  "low",
  "close",
  "binEdge",
  "binCount",
  "label",
]);
export type ChannelRole = z.infer<typeof channelRoleSchema>;

/**
 * One contiguous run of numbers inside the payload buffer.
 *
 * Channels are how we avoid ever putting bulk numeric data through JSON: the
 * descriptor stays small and human-readable, and every actual number lives in a
 * single flat byte buffer that decodes to a TypedArray with zero copying.
 */
export const channelSchema = z.object({
  name: z.string(),
  role: channelRoleSchema,
  dtype: wireDtypeSchema,
  length: z.number().int().nonnegative(),
  byteOffset: z.number().int().nonnegative(),
  byteLength: z.number().int().nonnegative(),
  stats: numericStatsSchema.nullable(),
});
export type Channel = z.infer<typeof channelSchema>;

/**
 * Records that the transferred payload is a reduction of the real value.
 *
 * Always surfaced in the UI. The user must never be left believing they are
 * looking at every point when they are not.
 */
export const decimationSchema = z.object({
  method: z.enum(["lttb", "minmax", "stride"]),
  originalLength: z.number().int().nonnegative(),
  outputLength: z.number().int().nonnegative(),
});
export type Decimation = z.infer<typeof decimationSchema>;

/**
 * The visible slice of the x axis, when the capture was restricted to one.
 *
 * Its statistics are separate from `Descriptor.stats` on purpose: those always
 * describe the complete value, and zooming must not quietly redefine them. The
 * property that makes them trustworthy is that they do not move with the view.
 */
export const windowSchema = z.object({
  from: z.number(),
  to: z.number(),
  stats: numericStatsSchema.nullable(),
});
export type WindowInfo = z.infer<typeof windowSchema>;

export const timeUnitSchema = z.enum(["s", "ms", "us", "ns"]);
export type TimeUnit = z.infer<typeof timeUnitSchema>;

export const indexInfoSchema = z.object({
  kind: z.enum([
    "range",
    "integer",
    "float",
    "datetime",
    "timedelta",
    "categorical",
    "multi",
    "other",
  ]),
  name: z.string().nullable(),
  /** Native dtype string, e.g. "datetime64[ns]". */
  dtype: z.string().nullable(),
  /** Name of the channel carrying index values, or null if it is implicit 0..n-1. */
  channel: z.string().nullable(),
  /** For datetime/timedelta indexes: the unit the channel's numbers are in. */
  timeUnit: timeUnitSchema.nullable(),
});
export type IndexInfo = z.infer<typeof indexInfoSchema>;

export const columnInfoSchema = z.object({
  name: z.string(),
  /** Native dtype string, e.g. "float64", "object", "category". */
  dtype: z.string(),
  /** Whether this column could be converted to a numeric channel. */
  numeric: z.boolean(),
  /** Name of the channel carrying this column's values, or null if not transferred. */
  channel: z.string().nullable(),
  stats: numericStatsSchema.nullable(),
});
export type ColumnInfo = z.infer<typeof columnInfoSchema>;

export const valueKindSchema = z.enum([
  "scalar",
  "ndarray",
  "series",
  "frame",
  "index",
  "sequence",
  "mapping",
  "object",
  "unsupported",
]);
export type ValueKind = z.infer<typeof valueKindSchema>;

export const vizKindSchema = z.enum([
  "line",
  "scatter",
  "histogram",
  "heatmap",
  "image",
  "grid",
  "tree",
  "scalar",
]);
export type VizKind = z.infer<typeof vizKindSchema>;

/**
 * Everything known about a captured value except its bulk numbers.
 *
 * Deliberately self-describing: the webview picks a visualization, labels axes
 * and renders the stats strip from this object alone, without knowing which
 * Python adapter produced it. That is what lets third-party adapters plug in
 * without touching the UI.
 */
export const descriptorSchema = z.object({
  kind: valueKindSchema,
  /** Fully qualified Python type, e.g. "numpy.ndarray", "pandas.core.frame.DataFrame". */
  pythonType: z.string(),
  /** Short, always-safe-to-display repr. Length-capped by the runtime. */
  preview: z.string(),
  shape: z.array(z.number().int().nonnegative()).nullable(),
  /** Native dtype string as Python sees it, e.g. "float64". */
  dtype: z.string().nullable(),
  /** In-memory size of the original value, not of the payload. */
  nbytes: z.number().int().nonnegative().nullable(),
  stats: numericStatsSchema.nullable(),
  index: indexInfoSchema.nullable(),
  columns: z.array(columnInfoSchema).nullable(),
  channels: z.array(channelSchema),
  decimation: decimationSchema.nullable(),
  /** Set when the capture covers only part of the x axis, after zooming. */
  window: windowSchema.nullable(),
  /** True when parts of the value were dropped entirely (e.g. columns beyond a cap). */
  truncated: z.boolean(),
  /** Adapter's ranked suggestions; the webview may override. */
  suggestedViz: z.array(vizKindSchema),
});
export type Descriptor = z.infer<typeof descriptorSchema>;
